tests/test_helper_functions.py

```diff
@@ -0,0 +1,170 @@
+    get_domain,
+    is_complete_url,
+    extract_url,
+    convert_http_to_https,
+)
+
+
+# Tests for the function `get_domain(df)`
+class TestGetDomain(unittest.TestCase):
+    def test_with_supported_schemes(self):
+        # Test with supported schemes and expect the correct domain
+        urls_with_expected_domains = [
+            ("http://example.com", "example.com"),
+            ("https://www.example.com", "www.example.com"),
+            ("git://github.com/user/repo.git", "github.com"),
+        ]
+
+        for url, expected in urls_with_expected_domains:
+            with self.subTest(url=url):
+                result = get_domain(url)
+                assert_that(result, is_(equal_to(expected)), f"Failed for URL: {url}")
+
+    def test_with_unsupported_schemes(self):
+        # Test with unsupported schemes and expect None
+        urls_with_unsupported_schemes = [
+            "ftp://example.com",
+            "file:///tmp/test",
+            "mailto:user@example.com",
+        ]
+
+        for url in urls_with_unsupported_schemes:
+            with self.subTest(url=url):
+                result = get_domain(url)
+                assert_that(result, is_(None), f"Failed for URL: {url}")
+
+    def test_with_malformed_urls(self):
+        malformed_urls = ["", "http:///example.com", None]
+        for url in malformed_urls:
+            with self.subTest(url=url):
+                result = get_domain(url)
+                assert_that(result, is_(equal_to(None)), f"Failed for URL: {url}")
+
+
+# Tests for the function `is_complete_url`
+EXPECTED_URL_PARTS = 5
+
+
+class TestIsCompleteURL(unittest.TestCase):
+    def test_complete_urls(self):
+        # Test with a URL that meets the expected part count
+        url = "https://github.com/owner/repo"
+        assert_that(is_complete_url(url), is_(True))
+
+    def test_incomplete_urls(self):
+        # Test with URLs that do not meet the expected part count
+        incomplete_urls = [
+            "https://github.com/owner",  # Only 4 parts
+            "http://github.com",  # Only 3 parts
+            "github.com/owner/repo",
+        ]  # Only 3 parts
+
+        for url in incomplete_urls:
+            with self.subTest(url=url):
+                assert_that(is_complete_url(url), is_(False))
+
+    def test_non_string_input(self):
+        # Test with non-string inputs. Expect the result to be False
+        non_strings = [123, None, [], {}]
+        for non_string in non_strings:
+            with self.subTest(input=non_string):
+                assert_that(is_complete_url(non_string), is_(False))
+
+    def test_with_edge_cases(self):
+        # Test with an empty string
+        assert_that(is_complete_url(""), is_(False))
+
+
+# Tests for the function `extract_url`
+class TestExtractURL(unittest.TestCase):
+    def test_empty_input(self):
+        result, flag = extract_url("")
+        assert_that(result, is_(None))
+        assert_that(flag, is_(True))
```

Comment:
The variable name `flag` doesn't communicate the purpose of the flag, so when reading the various `assert_that` statements here and elsewhere in these tests that test the value of `flag` don't read well. Please can you find a more descriptive name to replace `flag`?

Reply:
Change the variable name `flag` to `extraction_failed'.